### DISCUSSION TITLE
What is Mojo's approach to tail call optimization?

### DISCUSSION THREAD

**Discussion by Brian-M-J** [modular/modular#519] [Q&A] [answered]
Python famously does not have tail call optimization as Guido van Rossum [is against it](https://neopythonic.blogspot.com/2009/04/final-words-on-tail-calls.html). Since Mojo aims to be fully compatible with Python, what would its compiler do when it encountered an opportunity for this optimization? Would it check if the code is pure Python and then avoid performing TCO but perform it when it encounters Mojo-specific code? Would it provide a decorator like `always_tail_optimize` so the programmer could choose whether TCO is performed or not?

**Comment by lattner** [comment, marked as answer]
Yes, we will support tail call optimization and it will probably be enabled by default.  If needed, we can add a decorator to disable it though.

**Reply by Brian-M-J** [reply]
Hmm, what about also adding a decorator similar to [this Rust crate](https://docs.rs/tailcall/latest/tailcall/) that guarantees tail call optimizations? It would certainly be nice if programmers had compiler feedback on whether their tail recursive functions were implemented correctly.

There's also the concepts of [thunks](https://jtauber.com/blog/2008/03/30/thunks,_trampolines_and_continuation_passing/), [trampolines](https://pypi.org/project/trampoline/) and [continuation passing style](https://eli.thegreenplace.net/2017/on-recursion-continuations-and-trampolines/), which I honestly do not understand at all. Maybe people who do understand them can decide on whether these are features that should be added to the standard library or not.

**Comment by Brian-M-J** [comment]
> [thunks](https://jtauber.com/blog/2008/03/30/thunks,_trampolines_and_continuation_passing/), [trampolines](https://pypi.org/project/trampoline/) and [continuation passing style](https://eli.thegreenplace.net/2017/on-recursion-continuations-and-trampolines/), which I honestly do not understand at all

It's been about 8 months since I said that. I've learned a lot since then.

Cringing at myself suggesting thunks, trampolines and CPS be added as standard library features to a PL that already has first class functions aside, I have some more ideas for Mojo supporting tail call optimization.

Mojo already supports[^1] tail recursion optimization (tail calls to _the same_ function are replaced with jump instructions and reuse the stack frame), but [it's not clear](https://discord.com/channels/1087530497313357884/1206995505700147254/1207363538943746119) whether it supports tail call optimization (tail calls to _any_ function are replaced with jump instructions and reuse the stack frame). These may not be exactly the right terms but they convey the difference I'm trying to get across.

Generalizing tail call optimization to any function means that patterns like Continuation Passing Style[^2] and Finite State Machines[^3] can easily be expressed.

I'm also still convinced that it's better to syntactically guarantee the optimization so programmers don't have to second guess themselves (without this, an especially big footgun would be tail calls that implicitly use the `return` keyword i.e. tail calls to functions that return `None`). They also wouldn't have to write comments like:
```mojo
# I'm relying on TCO here. If you are modifying this function, make sure it still performs TCO or the stack will blow up
```
This can be done in two ways:
- Add a compiler intrinsic similar to clang's [musttail](https://clang.llvm.org/docs/AttributeReference.html#musttail):
This is something I'd rather avoid. Compiler intrinsics are completely foreign to Python and if we can avoid them, we should.
- Add a decorator to guarantee tail call optimizations:
The decorator, when placed on a function, guarantees that all `return` statements (even implicit ones) are tail calls. If not, compilation fails. This decorator can be thought of as automatically annotating all explicit and implicit `return` statements with `[[clang::musttail]]`. This is more restrictive than the choice to annotate individual `return` statements, but it is better for programmers' sanity that _all_ `return` statements are considered.
Ideally, using this decorator should require no manual transformations of the function by the programmer. If you look at the Python trampoline package, it requires that `return` statements be replaced by `yield`s. `[[clang::musttail]]` requires that the `return` keyword be used, even for calls to functions that return `void`. It's better to avoid making the programmer do extra work like this.[^4]
Adding a decorator also means leaving room for better support for TCO, like [TCO modulo cons](https://jfmengels.net/modulo-cons/) and [TCO modulo context](https://www.microsoft.com/en-us/research/publication/tail-recursion-modulo-context-an-equational-approach-2/)[^5]. Analysis can be performed on functions decorated as such and programmers don't need to maintain explicit accumulators.

I also hope Mojo's support for TCO isn't as restrictive as `musttail`'s:
> The target function must have the same number of arguments as the caller. The types of the return value and all arguments must be similar according to C++ rules (differing only in cv qualifiers or array size), including the implicit “this” argument, if any. Any variables in scope, including all arguments to the function and the return value must be trivially destructible. The calling convention of the caller and callee must match, and they must not be variadic functions

Additional Reading:
1. [WebAssembly tail calls](https://v8.dev/blog/wasm-tail-call)
2. [p2300 - std::execution](https://wg21.link/p2300):
> because all the work is done within callbacks invoked on the completion of an earlier sender, recursively up to the original source of computation, the compiler is able to see a chain of work described using senders as a tree of tail calls, allowing for inlining and removal of most of the sender machinery. In fact, when work is not submitted to execution resources outside of the current thread of execution, compilers are capable of removing the senders abstraction entirely, while still allowing for composition of functions across different parts of a program.
3. [WebAssembly proposal for tail call optimizations](https://github.com/WebAssembly/tail-call/blob/main/proposals/tail-call/Overview.md)
4. [The Story of Tail Call Optimizations in Rust](https://seanchen1991.github.io/posts/tco-story/)
5. Reviving tail-call elimination (issue 2691 at https://github.com/rust-lang/rfcs)
6. [Tail Call Optimization and Memory Management with Trait Objects and Vtables in Rust](https://www.eventhelix.com/rust/rust-to-assembly-tail-call-via-vtable-and-box-trait-free/)
7. [It’s not always obvious when tail-call optimization is allowed](https://quuxplusone.github.io/blog/2021/01/09/tail-call-optimization/)
8. [ES2015, ES2016, and beyond](https://v8.dev/blog/modern-javascript#proper-tail-calls)
9. C2 Wiki - [Tail Call Optimization](https://wiki.c2.com/?TailCallOptimization)
10. [Parsing Protobuf at 2+GB/s: How I Learned To Love Tail Calls in C](https://blog.reverberate.org/2021/04/21/musttail-efficient-interpreters.html)

[^1]: See the 'Tail Call Optimization (TCO)' section of [this](https://www.modular.com/blog/mojo-vs-rust-is-mojo-faster-than-rust) article.
[^2]: See [here](https://opendsa-server.cs.vt.edu/OpenDSA/Books/PL/html/FP9.html) and [here](https://matt.might.net/articles/by-example-continuation-passing-style/) for examples of what TCO and CPS enable programmers to do.
[^3]: See [here](https://www.lua.org/pil/6.3.html) for an example of an FSM implemented using TCO.
[^4]: The Rust trampoline crate doesn't have this problem, but it is restricted to TRO only.
[^5]: If video's more your style, you can watch [this](https://youtu.be/TbTdEW3Zkk0)
